bc_obps/service/form_builder_service.py

```diff
@@ -0,0 +1,196 @@
+        )
+        .first()
+        .json_schema
+    )
+    gas_type_map = {}
+    methodology_map = {}
+
+    # Fetch valid source_type(s) for the activity
+    valid_source_types = (
+        ConfigurationElement.objects.select_related('source_type')
+        .filter(
+            reporting_activity_id=activity, valid_from__valid_from__lte=report_date, valid_to__valid_to__gte=report_date
+        )
+        .distinct('source_type__id')
+    )
+
+    # Except if no valid source_types are found
+    if valid_source_types.count() == 0:
+        raise Exception(f'No valid source_types found for activity_id {activity} & report_date {report_date}')
+    # If an activity only has one source_type, the source type is mandatory and should be added to the schema
+    elif valid_source_types.count() == 1:
+        rjsf_schema['properties']['sourceTypes'] = {"type": "object", "title": "Source Types", "properties": {}}
+        build_source_type_schema(
+            activity, valid_source_types[0].source_type_id, report_date, rjsf_schema, gas_type_map, methodology_map
+        )
+
+    # If there are multiple source_types for an activity, the user may choose which ones apply. The IDs of the selected source_types are passed as a list in the parameters & we add those schemas to the activity schema.
+    else:
+        for s in valid_source_types:
+            rjsf_schema['properties'][str_to_camel_case(s.source_type.name)] = {
+                "type": "boolean",
+                "title": s.source_type.name,
+            }
+
+    # If no source_types are passed & there are more than 1 valid source type, only return the activity schema
+    if len(source_types) == 0 and valid_source_types.count() > 1:
+        return json.dumps(rjsf_schema)
+
+    if valid_source_types.count() > 1:
+        # Create the Source Types object (within which all the selected source_type schemas will be defined)
+        rjsf_schema['properties']['sourceTypes'] = {"type": "object", "title": "Source Types", "properties": {}}
+        # For each selected source_type, add the related schema
+        for source_type in source_types:
+            build_source_type_schema(activity, source_type, report_date, rjsf_schema, gas_type_map, methodology_map)
+
+    # Return completed schema
+    return_object = {}
+    return_object["schema"] = rjsf_schema
+    return_object["gasTypeMap"] = gas_type_map
+    return_object['methodologyMap'] = methodology_map
+    return json.dumps(return_object)
+
+
+class FormBuilderService:
+    @classmethod
+    def build_form_schema(request, activity=None, source_types=[], report_date='2024-04-01'):
+        if report_date is None:
+            raise Exception('Cannot build a schema without a valid report date')
+        if activity is None:
+            raise Exception('ERROR: Cannot build a schema without Activity data')
+        schema = build_schema(activity, source_types, report_date)
```

Comment:
before build_schema is called, we can get the configuration object here and pass it as a parameter, replacing report_date at this point with the set of configs that report_date applies to. We can then replace the above nested date comparisons like so:
`valid_from__valid_from__lte=report_date, valid_to__valid_to__gte=report_date`
->
`valid_from__lte=config.id, valid_to__gte=config.id`